### PR TITLE
Unbox across static handlers

### PR DESCRIPTION
**Pull request by alainfrisch**
This is a version of #2162, but built on top of the new (proposed) treatment of unboxing at the Cmm level (from #2165).

The first two commits are those from #2165.  Only the third commit should be reviewer here, and this PR should not be merged before #2165.

### CODE REVIEW

File: asmcomp/cmm.ml
Comment (lthls) [inline asmcomp/cmm.ml]: If I remember correctly, the fields of `Cifthenelse` are usually labelled `(cond, ifso_dbg, ifso, ifnot_dbg, ifnot, dbg)`. Ignoring the names themselves, this suggests that it's not clear which debug info correspond to which expression, and it should be made explicit in the type definition (either by grouping expressions with their associated debug info as a tuple, or using an inline record).
This is orthogonal to this pull request, of course.

Comment (gasche) [reply asmcomp/cmm.ml]: (Tuples have my preference, or any other mean of grouping related data together.)

Comment (mshinwell) [reply asmcomp/cmm.ml]: I have a patch to tidy up this type along the same lines as the `Lambda.block` construct under discussion elsewhere.

Comment (alainfrisch) [reply asmcomp/cmm.ml]: > This is orthogonal to this pull request, of course.

The choice of variable names was harmless but confusing.  Fixed!

**Comment by alainfrisch** [comment]
@lthls : would you like to review this one as well?

**Review by lthls — APPROVED** [review]
I approve this pull request, both its objective and its implementation. There are a few remarks that I can make, tough:

- If I understand correctly, after this gets merged we'll have catch handlers with non-`val` register types for the first time on trunk. I think this is already handled correctly, but I'll ask a few people around me who have experienced with register types if there's anything to watch for.

- The `notify_catch` field of the environment is a bit unintuitive. Knowing what it tries to achieve, I understand how it works, but it could use a few lines of comments explaining what it's used for. I think that the natural solution would be to have the `notify_catch` field contain an `unboxed_number_kind IntMap.t`, but this would mean that the environment contains information that needs to be returned back to the caller, changing the signature of the `transl*` functions and updating the code everywhere. I see the proposed version here as a way around that problem, but it could do with a bit more documentation.

- I'd be curious to know how often the `No_result` case occurs in `transl_catch`, and on which kind of programs. There's a number of optimisations that could be done if it happens, but I'd expect that either the optimisations are already done elsewhere or that this case never occurs in practice.

**Comment by alainfrisch** [comment]
Thanks @lthls.  I've added some words about the notify_catch callback technique.

>  I'll ask a few people around me who have experienced with register types if there's anything to watch for.

I think all the cmm and post-cmm backend is ready to deal with non-val types (also for functions arguments, which is what I used in an experimental PR about unboxed calling conventions), but indeed, this hasn't received wide coverage.  Did you more feedback from other people?

> I'd be curious to know how often the No_result case occurs in transl_catch, and on which kind of programs.

I haven't been able to produce this situation (corresponding to a dead catch handler), and possibly this is excluded by other passes, but it would seem weird to reject the case at this point.

**Comment by lthls** [comment]
> I think all the cmm and post-cmm backend is ready to deal with non-val types (also for functions arguments, which is what I used in an experimental PR about unboxed calling conventions), but indeed, this hasn't received wide coverage. Did you more feedback from other people?

Yes, they confirmed that they didn't find any bugs linked to this change in their experiments.

> I haven't been able to produce this situation (corresponding to a dead catch handler), and possibly this is excluded by other passes, but it would seem weird to reject the case at this point.

I wasn't thinking about rejecting the case, I would have simply removed the handler since it's unreachable. If you haven't encountered such a situation in practice, it's probably not worth worrying about.

Thanks for addressing my concerns.

In the light of the recently merged #8584, do you want to take the opportunity to show some of the improvements that one can expect ? Your original pull request (#2162) showed some of the cases where it would trigger, maybe you could build a micro-benchmark based on these.

**Comment by alainfrisch** [comment]
Sorry, I should indeed have put a link to https://github.com/ocaml/ocaml/pull/2165#issuecomment-442557042 which still applies (just checked).

**Comment by alainfrisch** [comment]
(That's a 40% speedup on the micro-benchmark.)

**Comment by alainfrisch** [comment]
Based on @lthls 's approval and the availability of the micro-benchmark above, I'll merge this soon if nobody complains.

**Comment by mshinwell** [comment]
@alainfrisch Have you obtained any benchmark numbers from larger programs with this patch?  It seems a significant enough change to be worth doing so.

As an aside, I think the existing register typing for unboxed integers is dubious, as they are assigned type `Int`.  However this could in theory be joined with `Val` to yield `Val`, which would be wrong for e.g. a boxed int64.  I suspect such a join never happens for a register holding a value that arose from unboxing, but this should probably be fixed nonetheless.  My previous PR on improved register typing, which has been taken over by @Gbury (in conjunction with work on register type inference, as discussed at the last dev meeting), should be sufficient.

**Comment by alainfrisch** [comment]
> @alainfrisch Have you obtained any benchmark numbers from larger programs with this patch? It seems a significant enough change to be worth doing so.

No visible change on some larger payloads I tried, but we wrote our critical numerical code knowing how unboxing works currently, and carefully rewrote parts to workaround current limitations.  This ends up with some code like:

````ocaml
  let a = ref 0. and b = 0. in
  if cond then (a := ...; b := ...) else (a:= ...; b := ...)
````

which could be avoided by a simple unboxed float tuple binding with this PR.

> I suspect such a join never happens for a register holding a value that arose from unboxing, but this should probably be fixed nonetheless.

If there is a problem, I think it already exists with normal unboxing of let-bound identifiers.  Here we unbox catch arguments if they are boxed numbers for sure, considering all call sites; and for the "continuation" (i.e. the handler code), it is as if we inserted a local let-bound identifier to hold the unboxed value (`let argi_unboxed = unbox argi in ...`).

**Comment by mshinwell** [comment]
I think the typing problem isn't specific to this PR, agreed.

**Comment by alainfrisch** [comment]
@mshinwell : are you ok with merging?

**Comment by mshinwell** [comment]
I think as long as you're satisfied with the potential performance changes, it's fine.

**Comment by alainfrisch** [comment]
Closed/reopened to relaunch Appveyor CI, which took significantly longer than usual (and one job was killed because of it); no problem on Travis, so I suspect a transient issue on Appveyor, but let's be sure before merging.

**Comment by alainfrisch** [comment]
Ok, AppVeyor finished each job in < 1 hour, but job time seems to be significantly higher than on master; I think it's worth investigating before merging.

**Comment by alainfrisch** [comment]
False alarm, the latest AppVeyor build is a fast as usual (also confirmed by checking AppVeyor logs that slow builds were also slow on the bytecode parts).

**Pull request merged**
